modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCdsProductPricer.java

```diff
@@ -0,0 +1,743 @@
+          bBar[i - 1] += dht / dhrt * (t0 + dt / dhrt);
+          bBar[i] += dht / dhrt * (-t1 - dt / dhrt);
+        }
+        t0 = t1;
+      }
+      pv += tPv;
+      ht0 = ht1;
+      rt0 = rt1;
+      b0 = b1;
+    }
+    double yfRatio = coupon.getYearFraction() /
+        discountFactors.getDayCount().relativeYearFraction(coupon.getStartDate(), coupon.getEndDate());
+    // pv sensitivity
+    PointSensitivityBuilder qSensiFirst = survivalProbabilities.zeroRatePointSensitivity(knots.get(0))
+        .multipliedBy(yfRatio * ((dhrtBar[0] + dhtBar[0]) / q[0] + bBar[0] * p[0]));
+    PointSensitivityBuilder pSensiFirst = discountFactors.zeroRatePointSensitivity(knots.get(0))
+        .multipliedBy(yfRatio * (dhrtBar[0] / p[0] + bBar[0] * q[0]));
+    PointSensitivityBuilder pvSensi = pSensiFirst.combinedWith(qSensiFirst);
+    for (int i = 1; i < nItems - 1; ++i) {
+      PointSensitivityBuilder qSensi = survivalProbabilities.zeroRatePointSensitivity(knots.get(i)).multipliedBy(
+          yfRatio * (-(dhrtBar[i - 1] + dhtBar[i - 1]) / q[i] + (dhrtBar[i] + dhtBar[i]) / q[i] + bBar[i] * p[i]));
+      PointSensitivityBuilder pSensi = discountFactors.zeroRatePointSensitivity(knots.get(i)).multipliedBy(
+          yfRatio * (-dhrtBar[i - 1] / p[i] + dhrtBar[i] / p[i] + bBar[i] * q[i]));
+      pvSensi = pvSensi.combinedWith(pSensi).combinedWith(qSensi);
+    }
+    if (nItems > 1) {
+      PointSensitivityBuilder qSensiLast = survivalProbabilities.zeroRatePointSensitivity(knots.get(nItems - 1)).multipliedBy(
+          yfRatio * (-(dhrtBar[nItems - 2] + dhtBar[nItems - 2]) / q[nItems - 1] + bBar[nItems - 1] * p[nItems - 1]));
+      PointSensitivityBuilder pSensiLast = discountFactors.zeroRatePointSensitivity(knots.get(nItems - 1)).multipliedBy(
+          yfRatio * (-dhrtBar[nItems - 2] / p[nItems - 1] + bBar[nItems - 1] * q[nItems - 1]));
+      pvSensi = pvSensi.combinedWith(pSensiLast).combinedWith(qSensiLast);
+    }
+
+    return Pair.of(yfRatio * pv, pvSensi);
+  }
+
+  //-------------------------------------------------------------------------
+  private double recoveryRate(ResolvedCds cds, CreditRatesProvider ratesProvider) {
+    RecoveryRates recoveryRates = ratesProvider.recoveryRates(cds.getLegalEntityId());
+    ArgChecker.isTrue(recoveryRates instanceof ConstantRecoveryRates, "recoveryRates must be ConstantRecoveryRates");
+    return recoveryRates.recoveryRate(cds.getProtectionEndDate());
+  }
+
+  private Pair<CreditDiscountFactors, LegalEntitySurvivalProbabilities> reduceDiscountFactors(
+      ResolvedCds cds, CreditRatesProvider ratesProvider) {
+    Currency currency = cds.getCurrency();
+    CreditDiscountFactors discountFactors = ratesProvider.discountFactors(currency);
+    ArgChecker.isTrue(discountFactors instanceof IsdaCompliantZeroRateDiscountFactors,
+        "discount factors must be IsdaCompliantZeroRateDiscountFactors");
```

Comment:
As `IsdaCompliantZeroRateDiscountFactors` is not being cast to, maybe it should be a method on `CreditDiscountFactors`? ie `isIsdaComplient()`? Not sure....